sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -505,20 +493,15 @@ public void processElement(ProcessContext c, BoundedWindow window) throws Except
         DestinationT destination = sink.getDynamicDestinations().getDestination(input);
         Writer<DestinationT, OutputT> writer = writers.get(destination);
         if (writer == null) {
-          LOG.debug("Opening writer for write operation {}", writeOperation);
+          String uuid = UUID.randomUUID().toString();
+          LOG.info(
+              "Opening writer {} for window {} pane {} destination {}",
+              uuid,
+              window,
+              c.pane(),
+              destination);
           writer = writeOperation.createWriter();
-          int shardNumber =
-              shardNumberAssignment == ShardAssignment.ASSIGN_WHEN_WRITING
```

Comment:
<!--thread_id:cc_152652760_t; commit:0df7ba9d6791bf06a59e3e7d0cb9500bbaa2b1ee; resolved:0-->
<!--section:context-quote-->
> **jkff** wrote:
> After https://github.com/apache/beam/pull/4124 they do - see https://github.com/apache/beam/pull/4137 for explanation.

<!--section:body-->
We went through some trouble to ensure that windowed writes in batch do not require fixed sharding. 4124 fixes a bug that only affects triggered (generally streaming) writes, so it's unfortunate if it makes the batch case less usable. Is there any way to make windowed writes work in batch without requiring fixed sharding?

Reply:
In #4137 I argued that this was never guaranteed to work. Or more like, it would only work if https://github.com/apache/beam/blob/80ae9303bd4c7326e18cd372da6abaf40435034a/sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java#L769 fires exactly once and deterministically contains all the data - the code used "collection is bounded" as a proxy for that, but in reality this is not sufficient: Beam model does not prevent having a bounded collection triggering multiple times.

However, it seems that the only reason it doesn't work is that the order of things retrieved from a GBK is unstable - well, let's stabilize it then, with another reshuffle! I did that, and now this is supported - thanks for pushing :) Done.